swarm/src/lib.rs

```diff
@@ -701,34 +714,31 @@ where TBehaviour: NetworkBehaviour<ProtocolsHandler = THandler>,
                     if this.banned_peers.contains(&peer_id) {
                         this.behaviour.inject_dial_failure(&peer_id);
                     } else {
-                        let result = match condition {
+                        let condition_matched = match condition {
                             DialPeerCondition::Disconnected
-                                if this.network.is_disconnected(&peer_id) =>
-                            {
-                                ExpandedSwarm::dial(this, &peer_id)
-                            }
+                                if this.network.is_disconnected(&peer_id) => true,
                             DialPeerCondition::NotDialing
-                                if !this.network.is_dialing(&peer_id) =>
-                            {
-                                ExpandedSwarm::dial(this, &peer_id)
+                                if !this.network.is_dialing(&peer_id) => true,
+                            _ => false
+                        };
+
+                        if condition_matched {
+                            if let Ok(true) = ExpandedSwarm::dial(this, &peer_id) {
+                                return Poll::Ready(SwarmEvent::Dialing(peer_id));
                             }
-                            _ => {
-                                log::trace!("Condition for new dialing attempt to {:?} not met: {:?}",
-                                    peer_id, condition);
-                                if let Some(mut peer) = this.network.peer(peer_id.clone()).into_dialing() {
+
+                        } else {
+                            log::trace!("Condition for new dialing attempt to {:?} not met: {:?}",
+                                peer_id, condition);
+                            match this.network.peer(peer_id.clone()) {
+                                Peer::Dialing(mut peer) => {
                                     let addrs = this.behaviour.addresses_of_peer(peer.id());
                                     peer.connection().add_addresses(addrs);
+                                },
+                                Peer::Connected(_) => {},
+                                Peer::Disconnected(..) | Peer::Local => {
+                                    this.behaviour.inject_dial_failure(&peer_id);
```

Comment:
I'm really not sure about calling `inject_dial_failure` when the dialing condition _was not met_. I would only expect a call to `inject_dial_failure` if my behaviour emitted `DialPeer` and the dialing condition was actually met, i.e. a new dialing attempt was actually initiated, for which I want to know about the result. In other woirds, I would expect `inject_dial_failure` for every failed dialing attempt, not if no attempt was made. Admittedly, as I wrote in https://github.com/libp2p/rust-libp2p/pull/1506, if a behaviour actually wants to know if the dialing condition is not met because it would otherwise still keep some state, another callback may be needed then.

Reply:
I guess it would be ok if we clearly document that `inject_dial_failure` is called also if the dialing  condition is not met, i.e. such that every `DialPeer` is guaranteed to be paired with `inject_connection_established` or `inject_dial_failure`, but then there should be no exception for when the peer is already connected, again because the behaviour may otherwise keep some state around.

Reply:
I can revert that.